### PR TITLE
feat(ui): add ClassRoom Timer

### PR DESCRIPTION
**Pull request by Matrixbirds**
1. Add `Timer` to `flat-components`.
2. Add `Timer` to `renderer-app` and `flat-web`.

### CODE REVIEW

File: desktop/renderer-app/src/pages/BigClassPage/index.tsx [hunk 1]
```diff
@@ -228,9 +229,12 @@ export const BigClassPage = observer<BigClassPageProps>(function BigClassPage()
...
                 <NetworkStatus networkQuality={classRoomStore.networkQuality} />
-                {!classRoomStore.isCreator && (
-                    <RoomInfo roomStatus={classRoomStore.roomStatus} roomType={RoomType.BigClass} />
-                )}
+                {!classRoomStore.isCreator ? (
+                    <RoomInfo
+                        roomStatus={classRoomStore.roomStatus}
+                        roomType={classRoomStore.roomInfo?.roomType}
+                    />
+                ) : classRoomStore.roomInfo?.beginTime && <Countdown state={classRoomStore.roomStatus !== RoomStatus.Started ? 'paused' : 'started'} beginTime={classRoomStore.roomInfo.beginTime} />}
```
Comment (BlackHole1) [inline desktop/renderer-app/src/pages/BigClassPage/index.tsx]: `<Countdown roomStatus={classRoomStore.roomStatus}`

File: packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    const h = Math.floor(seconds / 3600);
+    const m = Math.floor((seconds % 3600) / 60);
+    const s = Math.floor((seconds % 3600) % 60);
+    return (
+        (h > 0 ? String(h).padStart(2, "0") + ":" : "") +
+        String(m).padStart(2, "0") +
+        ":" +
+        String(s).padStart(2, "0")
+    );
+}
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx]: use data-fns function

```suggestion
const prefixInteger = (number: number): string => {
    return String(number).padStart(2, "0");
}

const renderTime = (seconds: number): string => {
    const { hours, minutes, seconds } = intervalToDuration({start: 0, end: seconds * 1000});
    const timeSuffix = `${prefixInteger(minutes)}:${prefixInteger(seconds)}`;

    return hours !== 0 ? ``${prefixInteger(hours)}:${timeSuffix}` : timeSuffix;
}
```

File: packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import React, { useRef, useState, useEffect, useCallback } from "react";
+import classnames from "classnames";
+import { useTranslation } from "react-i18next";
+import "./style.less";
+
+export type CountdownProps = {
+    state: "paused" | "started";
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx]:
```suggestion
    state: RoomStatus
```

File: packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+const useMounted = (): React.MutableRefObject<boolean> => {
+    const mounted = useRef<boolean>(false);
+    useEffect(() => {
+        mounted.current = true;
+        return () => {
+            mounted.current = false;
+        };
+    }, []);
+    return mounted;
+};
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Countdown/index.tsx]: use `useIsUnMounted`

see: https://github.com/netless-io/flat/blob/3edb21b682cebc62c45cc60d60d544d92e6afa70/desktop/renderer-app/src/utils/hooks/lifecycle.ts#L4-L13

**Review by BlackHole1 — CHANGES_REQUESTED** [review]
`Countdown` => `Timer`

File: desktop/renderer-app/src/components/AppStoreButton/index.tsx [hunk 5]
```diff
@@ -50,7 +50,7 @@ const apps: AddAppParams[] = [
 
 export const appIcons = {
     Monaco: codeEditorSVG,
-    Countdown: countdownSVG,
+    Timer: countdownSVG,
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/AppStoreButton/index.tsx]:
```suggestion
    Timer: timerSVG,
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    const timestamp = useClockTick(beginTime, 100, roomStatus);
+
+    const roomStatusCls = classnames(`countdown-${roomStatus}`);
+
+    const { t } = useTranslation();
+
+    return (
+        <span className="countdown-bar">
+            <span className={roomStatusCls}>{t("room-started")}</span>
+            <span>{format(timestamp, "mm:ss")}</span>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]: forget `hh`?

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        return stopTimer;
+    }, []);
+
+    return timestamp;
+};
+
+export const Timer: React.FC<TimerProps> = ({ roomStatus = RoomStatus.Paused, beginTime }) => {
+    const timestamp = useClockTick(beginTime, 100, roomStatus);
+
+    const roomStatusCls = classnames(`countdown-${roomStatus}`);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
    const roomStatusCls = classnames(`timer-${roomStatus}`);
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    const { t } = useTranslation();
+
+    return (
+        <span className="countdown-bar">
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
        <span className="timer-bar">
```

File: packages/flat-components/src/components/ClassroomPage/Timer/style.less [hunk 9]
```diff
@@ -0,0 +1,5 @@
+.countdown-bar {
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/style.less]:
```suggestion
.timer-bar {
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: desktop/renderer-app/src/pages/BigClassPage/index.tsx [hunk 10]
```diff
@@ -228,8 +229,18 @@ export const BigClassPage = observer<BigClassPageProps>(function BigClassPage()
...
+                        roomStatus={classRoomStore.roomStatus}
+                        roomType={classRoomStore.roomInfo?.roomType}
+                    />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline desktop/renderer-app/src/pages/BigClassPage/index.tsx]: FYI: https://eslint.org/docs/rules/no-negated-condition

I will add this rule to our project later.

```suggestion
                {classRoomStore.isCreator ? (
                    classRoomStore.roomInfo?.beginTime && (
                        <Timer
                            roomStatus={classRoomStore.roomStatus}
                            beginTime={classRoomStore.roomInfo.beginTime}
                        />
                    )
                ) : (
                    <RoomInfo
                        roomStatus={classRoomStore.roomStatus}
                        roomType={classRoomStore.roomInfo?.roomType}
                    />
```

File: desktop/renderer-app/src/pages/OneToOnePage/index.tsx [hunk 11]
```diff
@@ -186,8 +187,18 @@ export const OneToOnePage = observer<OneToOnePageProps>(function OneToOnePage()
...
+                        roomStatus={classRoomStore.roomStatus}
+                        roomType={classRoomStore.roomInfo?.roomType}
+                    />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline desktop/renderer-app/src/pages/OneToOnePage/index.tsx]: Ditto.

File: desktop/renderer-app/src/pages/SmallClassPage/index.tsx [hunk 12]
```diff
@@ -223,11 +224,18 @@ export const SmallClassPage = observer<SmallClassPageProps>(function SmallClassP
...
                         roomStatus={classRoomStore.roomStatus}
                         roomType={classRoomStore.roomInfo?.roomType}
                     />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline desktop/renderer-app/src/pages/SmallClassPage/index.tsx]: Ditto.

File: web/flat-web/src/pages/BigClassPage/index.tsx [hunk 13]
```diff
@@ -202,8 +203,18 @@ export const BigClassPage = observer<BigClassPageProps>(function BigClassPage()
...
+                        roomStatus={classRoomStore.roomStatus}
+                        roomType={classRoomStore.roomInfo?.roomType}
+                    />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline web/flat-web/src/pages/BigClassPage/index.tsx]: Ditto.

File: web/flat-web/src/pages/OneToOnePage/index.tsx [hunk 14]
```diff
@@ -162,8 +163,18 @@ export const OneToOnePage = observer<OneToOnePageProps>(function OneToOnePage()
...
+                        roomStatus={classRoomStore.roomStatus}
+                        roomType={classRoomStore.roomInfo?.roomType}
+                    />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline web/flat-web/src/pages/OneToOnePage/index.tsx]: Ditto.

File: web/flat-web/src/pages/SmallClassPage/index.tsx [hunk 15]
```diff
@@ -204,11 +205,18 @@ export const SmallClassPage = observer<SmallClassPageProps>(function SmallClassP
...
                         roomStatus={classRoomStore.roomStatus}
                         roomType={classRoomStore.roomInfo?.roomType}
                     />
+                ) : (
+                    classRoomStore.roomInfo?.beginTime && (
+                        <Timer
+                            roomStatus={classRoomStore.roomStatus}
+                            beginTime={classRoomStore.roomInfo.beginTime}
+                        />
+                    )
```
Comment (LitoMore) [inline web/flat-web/src/pages/SmallClassPage/index.tsx]: Ditto.

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: desktop/renderer-app/src/pages/BigClassPage/index.tsx (hunk 10 above)
Comment (Matrixbirds) [reply desktop/renderer-app/src/pages/BigClassPage/index.tsx]: Sure. Glad to see your config.

**Review by hyrious — DISMISSED** [review]

File: web/flat-web/src/components/AppStoreButton/index.tsx [hunk 16]
```diff
@@ -50,7 +50,7 @@ const apps: AddAppParams[] = [
 
 export const appIcons = {
     Monaco: codeEditorSVG,
-    Countdown: countdownSVG,
+    Timer: countdownSVG,
```
Comment (BlackHole1) [inline web/flat-web/src/components/AppStoreButton/index.tsx]:
```suggestion
    Timer: timerSVG,
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 17]
```diff
@@ -0,0 +1,75 @@
+import React, { useRef, useState, useEffect, useCallback } from "react";
+import classnames from "classnames";
+import { useTranslation } from "react-i18next";
+import "./style.less";
+import { useIsUnMounted } from "../../../utils/hooks";
+import { RoomStatus } from "../../../types/room";
+import { intervalToDuration } from "date-fns/esm";
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
import { intervalToDuration } from "date-fns/fp";
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+export type TimerProps = {
+    roomStatus: RoomStatus;
+    beginTime: number;
+};
+
+const paddingZero = (number?: number): string => {
+    return String(number).padStart(2, "0");
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx:15]:
```suggestion
    return number === undefined ? "00" : String(number).padStart(2, "0");
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+const useClockTick = (beginTime: number, delay: number, cbArgs: any): Duration => {
+    const [timestamp, updateTimestamp] = useState<number>(Date.now());
+
+    const updateTime = (state: string): void => {
+        if (state === RoomStatus.Started) {
+            updateTimestamp(Date.now());
+        }
+    };
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+    const args = useRef<typeof cbArgs | null>(cbArgs);
+
+    useEffect(() => {
+        args.current = cbArgs;
+    });
+
+    const startTimer = useCallback((): void => {
+        if (unmounted.current === false) {
+            updateTime(args.current);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
            if (state === RoomStatus.Started) {
                updateTimestamp(Date.now());
            }
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    const stopTimer = useCallback((): void => {
+        timer.current && window.clearTimeout(timer.current);
+    }, [timer]);
+
+    useEffect(() => {
+        startTimer();
+        return stopTimer;
+    }, []);
+
+    return intervalToDuration({ start: beginTime, end: timestamp });
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
    return intervalToDuration({
        start: beginTime,
        end: timestamp,
    });
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+export const Timer: React.FC<TimerProps> = ({ roomStatus = RoomStatus.Paused, beginTime }) => {
+    const timeDuration = useClockTick(beginTime, 100, roomStatus);
+
+    const roomStatusCls = classnames(`timer-${roomStatus}`);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+
+    const { t } = useTranslation();
+
+    return (
+        <span className="timer-bar">
+            <span className={roomStatusCls}>{t("room-started")}</span>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
            <span className={`timer-${roomStatus}`}>{t("room-started")}</span>
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+const paddingZero = (number?: number): string => {
+    return String(number).padStart(2, "0");
+};
+
+const renderTime = ({ hours, minutes, seconds }: Duration): string => {
+    return hours && hours > 0
+        ? `${paddingZero(hours)}:${paddingZero(minutes)}:${paddingZero(seconds)}`
+        : `${paddingZero(minutes)}:${paddingZero(seconds)}`;
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
    return hours > 0
        ? `${paddingZero(hours!)}:${paddingZero(minutes!)}:${paddingZero(seconds!)}`
        : `${paddingZero(minutes!)}:${paddingZero(seconds!)}`;
```

`hours` / `minutes` / `seconds`  must be `number`. see: https://github.com/date-fns/date-fns/blob/54ec0a5e5ca3f472085d3b56be30d801b90829ac/src/intervalToDuration/index.ts#L38

this is `date-fns` bug

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+import { useIsUnMounted } from "../../../utils/hooks";
+import { RoomStatus } from "../../../types/room";
+import { intervalToDuration } from "date-fns/esm";
+
+export type TimerProps = {
+    roomStatus: RoomStatus;
+    beginTime: number;
+};
+
+const paddingZero = (number?: number): string => {
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
const paddingZero = (number: number): string => {
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx (hunk 24 above)
Comment (BlackHole1) [reply packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]: see: https://github.com/date-fns/date-fns/pull/2917

Comment (BlackHole1) [reply packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]: https://github.com/date-fns/date-fns/issues/2891#issuecomment-1003070337

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+import { useIsUnMounted } from "../../../utils/hooks";
+import { RoomStatus } from "../../../types/room";
+import { intervalToDuration } from "date-fns/fp";
+
+export type TimerProps = {
+    roomStatus: RoomStatus;
+    beginTime: number;
+};
+
+export type TimerDuration = Required<Duration>;
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
// https://github.com/date-fns/date-fns/issues/2891#issuecomment-1003070337
export type TimerDuration = Omit<Required<Duration>, "weeks">;
```

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+
+const paddingZero = (number: number): string => {
+    return String(number).padStart(2, "0");
+};
+
+const renderTime = ({ hours, minutes, seconds }: TimerDuration): string => {
+    return hours > 0
+        ? `${paddingZero(hours)}:${paddingZero(minutes)}:${paddingZero(seconds)}`
+        : `${paddingZero(minutes)}:${paddingZero(seconds)}`;
+};
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
    const minutesAndSeconds = `${paddingZero(minutes)}:${paddingZero(seconds)};

    return hours > 0 ? `${paddingZero(hours)}:${minutesAndSeconds}` : minutesAndSeconds;
};
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+
+// see: https://github.com/date-fns/date-fns/issues/2891#issuecomment-1003070337
+export type TimerDuration = Omit<Required<Duration>, "weeks">;
+
+const paddingZero = (number: number): string => {
+    return String(number).padStart(2, "0");
+};
+
+const renderTime = ({ hours, minutes, seconds }: TimerDuration): string => {
+    const minutesAndSeconds = `${paddingZero(minutes)}:${paddingZero(seconds)};
```
Comment (BlackHole1) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]:
```suggestion
    const minutesAndSeconds = `${paddingZero(minutes)}:${paddingZero(seconds)}`;
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/src/pages/SmallClassPage/index.tsx [hunk 29]
```diff
@@ -44,6 +44,8 @@ import { ShareScreen } from "../../components/ShareScreen";
 import { generateAvatar } from "../../utils/generate-avatar";
 import { AppStoreButton } from "../../components/AppStoreButton";
 
+const a = "";
```
Comment (Cheerego7) [inline web/flat-web/src/pages/SmallClassPage/index.tsx]: what's this?

File: desktop/renderer-app/src/components/AppStoreButton/index.tsx [hunk 30]
```diff
@@ -1,7 +1,7 @@
 import "./style.less";
 import appStoreSVG from "../../assets/image/app-store.svg";
 import codeEditorSVG from "../../assets/image/code-editor.svg";
-import countdownSVG from "../../assets/image/countdown.svg";
+import countdownSVG from "../../assets/image/countdonw.svg";
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/AppStoreButton/index.tsx]:
```suggestion
import countdownSVG from "../../assets/image/countdown.svg";
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/src/pages/SmallClassPage/index.tsx (hunk 29 above)
Comment (BlackHole1) [reply web/flat-web/src/pages/SmallClassPage/index.tsx]: this is test code

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+const useClockTick = (beginTime: number, delay: number, roomStatus: RoomStatus): TimerDuration => {
+    const [timestamp, updateTimestamp] = useState<number>(Date.now());
+
+    const unmounted = useIsUnMounted();
+
+    const timer = useRef<number>(NaN);
+
+    const state = useRef<typeof roomStatus>(roomStatus);
```
Comment (Cheerego7) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]: `state` -> `roomState`

File: packages/flat-components/src/components/ClassroomPage/Timer/index.tsx [hunk 32]
```diff
@@ -0,0 +1,60 @@
+import React, { useState, useEffect, useMemo } from "react";
+import { useTranslation } from "react-i18next";
+import "./style.less";
```
Comment (Cheerego7) [inline packages/flat-components/src/components/ClassroomPage/Timer/index.tsx]: placed at the top, and add a newline

**Review by BlackHole1 — APPROVED** [review]

**Comment by BlackHole1** [comment]
please re-review

cc: @LitoMore

**Review by Cheerego7 — APPROVED** [review]

**Review by LitoMore — APPROVED** [review]

**Review by LitoMore — APPROVED** [review]

**Pull request merged**